Shapes - Alternate

Redth opened the pull request:
This builds on and is an alternate approach to #625 and combines the existing Shape implementations (Ellipse, Rectangle, Line, etc) with the new IShape implementations.

### PR Checklist ###

- [x] Targets the correct branch 
- [ ] Tests are passing (or failures are unrelated)
- [ ] Adds the property to the appropriate interface
- [ ] Avoids any changes not essential to the handler property
- [ ] Adds the mapping to the PropertyMapper in the handler
- [ ] Adds the mapping method to the Android, iOS, and Standard aspects of the handler
- [ ] Implements the actual property updates (usually in extension methods in the Platform section of Core)
- [ ] Tags ported renderer methods with [PortHandler]
- [ ] Adds an example of the property to the sample project (MainPage)
- [ ] Adds the property to the stub class
- [ ] Implements basic property tests in DeviceTests

#### Does this PR touch anything that might affect accessibility?
- [ ] Does this PR introduce a new control? (If yes, add an example using SemanticProperties to the SemanticsPage)
- [ ] APIs that modify focusability?
- [ ] APIs that modify any text property on a control?
- [ ] Does this PR modify view nesting or view arrangement in anyway?
- [ ] Is there the smallest possibility that your PR will change accessibility? 
- [ ] I'm not sure, please help me

If any of the above checkboxes apply to your PR, then the PR will need to provide testing to demonstrate that accessibility still works.

jsuarezruiz committed:
Implement Shapes using Microsoft.Maui.Graphics

jsuarezruiz committed:
Added ShapesExtensions WinUI class

jsuarezruiz committed:
Updated IShape

jsuarezruiz committed:
Updated Maui Graphics versions

jsuarezruiz committed:
Implement on Windows

jsuarezruiz committed:
Added BindableProperties to ShapeView

Redth committed:
Shape doesn't need density

Redth committed:
Make Geometry implement IShape

Redth committed:
Collapse shape impl's into old shape api's

Redth commented on src/Controls/src/Core/Shapes/Ellipse.cs:
```diff
@@ -1,11 +1,22 @@
+
+			path.AppendEllipse(rect);
+
+			return path.AsScaledPath((float)Width / (float)rect.Width);
```
This is probably unnecessary for shapes which are 'generated' by using a given rectangle.

Redth commented on src/Controls/src/Core/Shapes/GeometryGroup.cs:
```diff
@@ -97,5 +97,17 @@ void Invalidate()
+			var path = new Graphics.PathF();
+			foreach (var c in Children)
+			{
+				// TODO: Combine c.PathForBounds(rect) with main path
```
We need PathF to support concatenating multiple paths into one.  @jonlipsky ^

Redth commented on src/Controls/src/Core/Shapes/Line.cs:
```diff
@@ -38,5 +52,15 @@ public double Y2
+			path.MoveTo((float)X1, (float)Y1);
+			path.LineTo((float)X2, (float)Y2);
+
+			return path.AsScaledPath((float)Width / (float)rect.Width);
```
Since Line has a Width, and start/end points are given for specific values, we need to assume the Path is created in the bounds of the View of Line, and then scale it appropriately for the given rectangle.  Is this correct?

Redth commented on src/Controls/src/Core/Shapes/LineGeometry.cs:
```diff
@@ -32,5 +32,13 @@ public Point EndPoint
+			var path = new PathF();
+			path.Move((float)StartPoint.X, (float)StartPoint.Y);
+			path.LineTo((float)EndPoint.X, (float)EndPoint.Y);
+			return path;
```
Geometry doesn't have a concept of bounds before being asked for the path.  This is kind of strange, so we just have to basically assume the given fixed values are correct?  Should we do any bounds checking here?

Redth commented on src/Controls/src/Core/Shapes/Path.cs:
```diff
@@ -1,8 +1,26 @@
+		public PathF PathForBounds(Graphics.Rectangle rect)
+		{
+			return Data.PathForBounds(new Graphics.Rectangle(0, 0, Width, Height))
+				.AsScaledPath((float)Width / (float)rect.Width);
```
Again, since the Path has a Width/Height, we have to assume the path is for a given set of bounds, and we scale it for the given rectangle.

Redth commented on src/Controls/src/Core/Shapes/PathGeometry.cs:
```diff
@@ -37,5 +39,108 @@ public FillRule FillRule
+
+		public override Graphics.PathF PathForBounds(Graphics.Rectangle rect)
+		{
+			var path = new Graphics.PathF();
```
How should we handle Fill Rules here? PathF does not support fill rules.

Redth commented on src/Controls/src/Core/Shapes/Rectangle.cs:
```diff
@@ -25,5 +27,14 @@ public double RadiusY
+		{
+			var path = new PathF();
+
+			path.AppendRoundedRectangle(rect, (float)RadiusY + (float)RadiusX);
```
Need to add the actual path building math for dealing with RadiusY and RadiusX

Redth committed:
Scale ellipse for the view bounds

Redth committed:
Use parameters for ellipse geometry

Redth commented on src/Controls/src/Core/Shapes/Geometry.cs:
```diff
@@ -1,7 +1,9 @@
+	public abstract class Geometry : BindableObject, IShape
 	{
-
+		public abstract PathF PathForBounds(Graphics.Rectangle rect);
```
I'm not convinced this is the right approach.  Consider a Line with absolute Start/End Points.  Since the Geometry itself has no concept of a width/height, there's no way to know how to scale the start/end points and the line itself to the requested rectangle to return the path for.

In practice, on Geometry, PathForBounds is always passed the View's bounds that will display the Geometry so this should _work_ but it's rather confusing.

Maybe a `public abstract void AppendPath(PathF path)` would be more descriptive.

Redth replied:
Switched to this.

Redth committed:
Make Geometry append to path passed in

Geometry has no concept of width/height so it makes no sense to expect it to scale its path to a given bounds.  Instead, remove IShape and make `AppendToPath` and allow the geometry to just append its path operations to the existing path.

Redth committed:
Remove comment

jsuarezruiz committed:
Fix build errors

jsuarezruiz committed:
Fixed shapes handler tests

jsuarezruiz commented:
I have splitted into different classes the implementation of .NET MAUI to draw the Shape in Line.Impl.cs etc. classes and I have applied small changes to fix the Build.

The only question that I have from this approach is ... that by reusing Shapes from Xamarin.Forms we have a dilemma to solve. Shapes in Xamarin.Forms are Views, so why would we need ShapeView if each Shape is already a View?

Redth committed:
Remove bounds from shape path request

Redth committed:
Rename append method

Redth committed:
Add aspect to IShapeView

mattleibow commented on src/Core/tests/DeviceTests/Stubs/EllipseStub.cs:
```diff
@@ -0,0 +1,16 @@
+
+			path.AppendEllipse(0f, 0f, (float)Width, (float)Height);
+
+			return path.AsScaledPath((float)Width / (float)rect.Width);
```
Looking at this reminds me of the old code. Maybe these methods should be helper methods in Core? Like PathHelper.CreateEllipse and pass in the relevant props. This way we can test the logic in the actual implementation. Or, do we even need all the shapes in the tests since they are not serving any purpose? We are not testing the logic in Controls, so why even have this shape?

I should merge my device tests for compat/controls PR...

Redth commented on src/Controls/src/Core/HandlerImpl/Rectangle.Impl.cs:
```diff
@@ -0,0 +1,16 @@
+		{
+			var path = new PathF();
+
+			path.AppendRoundedRectangle(0f, 0f, (float)Width, (float)Height, (float)RadiusY + (float)RadiusX);
```
For now this is wrong anyway since the old behaviour for Rectangle is different. Needs some more math to curve for the x and y radius values. I was just shortcutting for now.

Redth commented on src/Controls/src/Core/HandlerImpl/ShapeView.cs:
```diff
@@ -0,0 +1,108 @@
+			BindableProperty.Create(nameof(Aspect), typeof(Aspect), typeof(ShapeView), null);
+		
+		public static readonly BindableProperty FillProperty =
+			BindableProperty.Create(nameof(Fill), typeof(Paint), typeof(ShapeView), null);
```
Ahh yeah i think you’re right here

davidbuckleyni commented:
Are we having default sizes like x1 x2 cricle sqaure rectangle as default shape choices where u can place circle everywhere's

Redth committed:
Move PointCollection back to Controls

Redth committed:
Add implicit conversion to DoubleCollection

Redth committed:
Introduce new aspect enum for shapes

Redth committed:
Add extension to make Paint -> Color

Redth committed:
Make Shape an IShapeView

This removes the need for a separate ShapeView in controls.

Redth committed:
Fix sample to use shapes directly

Redth committed:
Make Stroke a Paint

Redth committed:
Fix ellipse path generation

Redth committed:
Fix up casts

Redth committed:
Bring back sample usage

Redth committed:
Add known converter

Redth committed:
Move back to PathForBounds and move scaling to Controls

Clipping will need the bounds specified when asking for a Path.  This reworks the Shape in Controls to move the scaling of the path up into Shape.  The scaling here will check the requested bounds compared to the view's bounds and scale accordingly (taking into account Stretch/Aspect).

Redth commented on src/Core/src/Hosting/AppHostBuilderExtensions.cs:
```diff
@@ -5,13 +5,37 @@
 {
 	public static partial class AppHostBuilderExtensions
 	{
+		static readonly Dictionary<Type, Type> DefaultMauiHandlers = new Dictionary<Type, Type>
```
@PureWeen  is this a bad rebase?

Redth commented on src/Compatibility/Core/src/AppHostBuilderExtensions.cs:
```diff
@@ -57,6 +57,12 @@ class FormsCompatBuilder : IMauiServiceBuilder
 			typeof(Stepper),
 			typeof(TimePicker),
 			typeof(Shell),
+			typeof(Shapes.Ellipse),
```
@PureWeen rebased and this isn't used anymore, can you help wire this up again?

Redth committed:
Add shapes to default handlers

jsuarezruiz committed:
Take StrokeThickness into account when drawing Shapes

jsuarezruiz committed:
Improved drawing logic considering StrokeThickness

Redth commented on src/Controls/src/Core/HandlerImpl/ShapeView.cs:
```diff
@@ -0,0 +1,108 @@
+			BindableProperty.Create(nameof(Fill), typeof(Paint), typeof(ShapeView), null);
+
+		public static readonly BindableProperty StrokeProperty =
+			BindableProperty.Create(nameof(Stroke), typeof(Color), typeof(ShapeView), null);
```
Eventually, but right now the graphics canvas only accepts color here.

Redth commented on src/Controls/src/Core/HandlerImpl/ShapeView.cs:
```diff
@@ -0,0 +1,108 @@
+
+namespace Microsoft.Maui.Controls.Shapes
+{
+	public class ShapeView : View, IShapeView
```
Gone!

jsuarezruiz committed:
Rollback an unnecessary change

jsuarezruiz committed:
Updated shapes handler tests

jsuarezruiz committed:
Remove (for now) the Aspect Shape code

jsuarezruiz committed:
Update sample
